pkg/mimirpb/compat_test.go

```diff
@@ -150,3 +150,93 @@ func TestFromPointsToSamples(t *testing.T) {
 
 	assert.Equal(t, expected, FromPointsToSamples(input))
 }
+
+func TestPreallocatingMetric(t *testing.T) {
+	t.Run("should be unmarshallable from the bytes of a default Metric", func(t *testing.T) {
+		metric := Metric{
+			Labels: []LabelAdapter{
+				{Name: "l1", Value: "v1"},
+				{Name: "l2", Value: "v2"},
+				{Name: "l3", Value: "v3"},
+				{Name: "l4", Value: "v4"},
+			},
+		}
+
+		metricBytes, err := metric.Marshal()
+		require.NoError(t, err)
+
+		preallocMetric := &PreallocatingMetric{}
+		require.NoError(t, preallocMetric.Unmarshal(metricBytes))
+
+		assert.Equal(t, metric, preallocMetric.Metric)
+	})
+
+	t.Run("should not break with invalid protobuf bytes (no panic)", func(t *testing.T) {
+		preallocMetric := &PreallocatingMetric{}
+		assert.Error(t, preallocMetric.Unmarshal([]byte{1, 3, 5, 6, 238, 55, 135}))
+	})
+
+	t.Run("should correctly preallocate Labels slice", func(t *testing.T) {
+		metric := Metric{
+			Labels: []LabelAdapter{
+				{Name: "l1", Value: "v1"},
+				{Name: "l2", Value: "v2"},
+				{Name: "l3", Value: "v3"},
+				{Name: "l4", Value: "v4"},
+				{Name: "l5", Value: "v5"},
+			},
+		}
+
+		metricBytes, err := metric.Marshal()
+		require.NoError(t, err)
+
+		preallocMetric := &PreallocatingMetric{}
+		require.NoError(t, preallocMetric.Unmarshal(metricBytes))
+
+		assert.Equal(t, metric, preallocMetric.Metric)
+		assert.Equal(t, len(metric.Labels), cap(preallocMetric.Labels))
```

Comment:
Nit: you can use `testing.AllocsPerRun` to verify the number of allocations (which should be 10 + 1 I guess)

Reply:
i tried that and it didn't work, that's why i decided to not include the test. 

the problem was that `Metric.Unmarshal` was doing the same number of allocations - 1. So it seemed like `PrealloctingMetric had no effect in terms of allocations. I didn't look into why further